bpo-28684: asyncio tests now handle PermissionError raised when bindi…

xdegaye opened the pull request:
…ng a unix socket


<!-- issue-number: bpo-28684 -->
https://bugs.python.org/issue28684
<!-- /issue-number -->

xdegaye committed:
bpo-28684: asyncio tests now handle PermissionError raised when binding a unix socket

1st1 reviewed (changes requested):
I would prefer to use `@unittest.skipIf(android)` approach instead.  `PermissionError` can occur on a normal Linux machine as a result of some regression in asyncio.  I don't want such cases to slip through our CI just because it sometimes happens on Android.

xdegaye commented:
@1st1 I understand your concern. Instead of ``@unittest.skipIf(android)`` I think it would be better to use a new decorator ``@support.skip_unless_bind_unix_socket`` to widen the scope of the changes:
* The root user on Android does not raise PermissionError, so it should be possible to run the asyncio test suite in that case without skipping those tests.
* Android runs now SElinux in enforcing mode and one of its policies restricts the access to unix sockets [1], so the problem is not Android specific and may impact other linux platforms depending on the way they are configured. As a confirmation of the fact that SElinux is involved, the remote command ``adb logcat`` prints this kind of AVC SElinux record whenever one of the asyncio tests fails with PermissionError:
```
11-23 15:36:11.859  2919  2919 W python  : type=1400 audit(0.0:20): avc: denied { create }
for name="tmpug8hwp0r" scontext=u:r:shell:s0 tcontext=u:object_r:shell_data_file:s0
tclass=sock_file permissive=0
```

[1] An interesting academic paper [The Misuse of Android Unix Domain Sockets and Security Implications](https://dl.acm.org/citation.cfm?id=2978297).

1st1 commented:
SGTM, let's go for `@support.skip_unless_bind_unix_socket`

xdegaye committed:
Use instead a new skip_unless_bind_unix_socket() decorator

The test.support.skip_unless_bind_unix_socket() decorator is used to
skip asyncio tests that fail because the platform lacks a functional
bind() function for unix domain sockets (as it is the case for non root
users on the recent Android versions that run now SELinux in enforcing
mode).

xdegaye committed:
Merge remote-tracking branch 'origin/master' into bpo-28684

xdegaye committed:
The decorator must be applied last

xdegaye committed:
Only one decorator is needed

1st1 commented on Lib/test/support/__init__.py:
```diff
@@ -2432,6 +2433,28 @@ def skip_unless_xattr(test):
+        with socket.socket(socket.AF_UNIX) as sock:
+            try:
+                sock.bind(path)
+                _bind_nix_socket_error = ''
```
`_bind_nix_socket_error = False`

xdegaye replied:
Yes, much better.

xdegaye committed:
Set _bind_nix_socket_error to False when no error

1st1 commented:
LGTM

xdegaye commented:
I am not sure that these changes need to be backported to 3.6.
